ibis/expr/types/joins.py

```diff
@@ -408,6 +409,68 @@ def asof_join(
         # return with a new JoinExpr wrapping the new join chain
         return self.__class__(left, collisions=collisions, equalities=equalities)
 
+    @functools.wraps(Table.temporal_join)
+    def temporal_join(
+        self: Table,
+        right: Table,
+        predicates=(),
+        *,
+        at_time: Column = None,
+        lname: str = "",
+        rname: str = "{name}_right",
+    ):
+        if not isinstance(right.op(), ops.VersionedDatabaseTable):
+            raise IbisError(
+                "Right-table is not versioned. "
+                "Temporal join is defined only when the right-table is versioned."
+            )
+        elif at_time is None and right.at_time is None:
```

Comment:
Is this just a convenience to support defining `at_time` for both the table and the join operation?

What I am thinking is to restrict the API here so that the `ops.VersionedDatabaseTable`'s `at_time` attribute become mandatory and we remove the `at_time` argument here, something like:

```py
ordinary = con.table("my_ordinary_table")
versioned = con.table("my_versioned_table")

ordinary.temporal_join(
    versioned.at_time(ordinary.timestamp_col),
    [...additional predicates...]
)
```

That way we could remove the `ops.TemporalJoinLink` node as well, since it wouldn't hold an extra `at_time` property since that is always defined on the right table (`how` would carry the information about the join kind).

Reply:
Actually the nicest would probably be your example from #8247:

```py
expr = table_left.join(
    table_right.at_time(table_left.event_time_attribute),
    table_left.on_field == table_right.on_field
)
```

If the right table passed to `.join()` is a versioned table, then the join must be a temporal join.

Reply:
Thanks. As you raised this earlier, I also like this API (API-2) more compared to the first one (API-1) requiring the `at_time` argument in `temporal_join()`. I have kept both API's so far in case one would require a more complex implementation than the other and that might influence our opinion about which one to pick. This did not turn out to be the case, so I will remove the option with `at_time` argument.

> If the right table passed to .join() is a versioned table, then the join must be a temporal join.

This is a good idea, will make this change.

There are, however, a few things that are currently not ideal with the implementation of API-2. I documented these items in `TODO`'s. Here is a summary of them:

- The current implementation requires `table_right.at_time(...)` to be on a separate line as
```
        table_right = table_right.at_time(table_left.timestamp_col)
        expr = table_left.temporal_join(
            table_right,
            predicates=[
                table_left["id"] == table_right["id"],
            ],
        )
```
If we do not over-write `table_right` as given above, i.e.,
```
        expr = table_left.temporal_join(
            table_right.at_time(table_left.timestamp_col),
            predicates=[
                table_left["id"] == table_right["id"],
            ],
        )
```
we will get
```
        E ibis.common.exceptions.IntegrityError: Cannot add <ibis.expr.operations.logical.Equals object at 0x17c374270> to projection, they belong to another relation
```
This is because `at_time()` returns a new copy of `table_right` and the predicates are defined with the old copy of `table_right`. 

- On removing `ops.TemporalJoinLink`:

> That way we could remove the ops.TemporalJoinLink node as well, since it wouldn't hold an extra at_time property since that is always defined on the right table (how would carry the information about the join kind).

I tried to implement this exact idea previously, and access `at_time` through `join_op -> table_right_op -> at_time_op` -- `join_sql()` in `dialects.py` is actually a deprecated-leftover from this effort and kept it for reference during the reviews. 
We first need to rebase `table_right.at_time` on `join_table_right` with
```
        # Rebase `at_time` on the JoinTable wrapping up the DatabaseTable
        # `at_time` field is based on.
        deref_left = dereference_mapping_left(left)
        at_time = at_time_op.replace(deref_left, filter=ops.Value)
```
Then to keep a reference to the rebased `at_time`, we need to make it an attribute of `right`. The only way that I could come up to do this was
``` 
right = right.op().copy(at_time=at_time)
```
However, this lead to the same error given above
```
        E ibis.common.exceptions.IntegrityError: Cannot add <ibis.expr.operations.logical.Equals object at 0x17c374270> to projection, they belong to another relation
```
This is again because now the (old) `right` table referenced in the predicate is different than the new `right` table. As far as I understand it, to get this working, we would need to have an additional step where we update the relevant `predicates` with the new `right` before calling `prepare_predicates()`.

@kszucs Do you have suggestions or ideas on addressing these two points?

Reply:
I just "discovered" that Flink supports two different kinds of temporal join with `LEFT` and `INNER`. I did not read this in the docs, just found out by trying it out. With this, `how` in `TemporalJoinLink` would be set to either `left` or `right`. This might be another reason to keep `TemporalJoinLink` rather than "encoding" the join type in `JoinLink` as `temporal-left` or `temporal-inner`.